#### 1. 자료구조와 배열

In [ ]:
# 학생 5명의 시험 점수를 입력받아 합계와 평균을 출력하기

print('학생 그룹 점수의 합계와 평균을 구합니다.')

score1 = int(input('1번 학생의 점수를 입력하세요.: '))
score2 = int(input('2번 학생의 점수를 입력하세요.: '))
score3 = int(input('3번 학생의 점수를 입력하세요.: '))
score4 = int(input('4번 학생의 점수를 입력하세요.: '))
score5 = int(input('5번 학생의 점수를 입력하세요.: '))

total = 0
total += score1
total += score2
total += score3
total += score4
total += score5

print(f'합계는 {total}점입니다.')
print(f'평균은 {total / 5}점입니다.')

- 배열의 장점
    - 생성할 때 원소의 개수를 자유롭게 지정할 수 있다
    - 특정 값에 접근하기 쉽다(Indexing)
    - 오름차순, 내림차순으로 정렬 할 수 있다

##### 리스트
- 원소를 변경할 수 있는(mutable) 객체

In [1]:
# 원소가 5개이면서 원소값이 없는 리스트 생성
list = [None] * 5
print(list)

[None, None, None, None, None]


##### 튜플
- 원소에 순서를 매겨 결합한 것으로 원소를 변경할 수 없는(immutable) 자료형

##### 리스트와 튜플 풀어내기
- Unpack : 좌변에는 여러 개의 변수를 두고 우변에는 리스트나 튜플을 놓으면 리스트나 튜플의 원소들이 순서대로 변수에 대입

##### 인덱스로 원소에 접근하기

In [2]:
x = [11, 22, 33, 44, 55, 66, 77]

x[0] # 11
x[-1] # 77
x[-4] = 3.14
x

[11, 22, 33, 3.14, 55, 66, 77]

- 인덱스식을 좌변에, 대입할 값을 우변에 두면 x[-4]에 있던 44(int형)에서 3.14(float형)로 자동 형 변환이 된다.
- 이 때는 값 자체가 복사되는 것이 아니라 x[-4]가 참조하는 객체의 식별 번호가 변경

##### 슬라이스
- s[i:j] : s[i]부터 s[j-1]까지 나열
- s[i:j:k] : s[i]부터 s[j-1]까지 k씩 건너뛰며 나열
- i, j가 len(s)보다 크면 len(s)가 지정된 것으로 간주
- i가 없거나 None이면 0이 지정된 것
- j가 없거나 None이면 len(s)가 지정된 것
- s[:] : s의 모든 원소를 출력
- s[-n:] : s의 원소 중 뒤에서 n번째 원소부터 맨 뒤의 원소까지 출력
- s[::-1] : s의 모든 원소를 거꾸로 출력

##### mutable과 immutable의 대입
```python
n = 5
id(n) # 객체의 식별 번호
n = 'ABC'
id(n) # 객체의 식별 번호가 바뀜
```
- 변수에 어떤 값을 대입하면 값이 아니라 식별 번호가 바뀐다
- 이 원리로 어떤 자료형의 객체이든 상관없이 변수에 대입 가능
```python
x = 6
y = 2
x, y = y + 2, x + 3  # 두 개의 값이 동시에 업데이트됨
print(x) # 2 + 2 = 4
print(y) # 6 + 3 = 9
```
x = y + 2, y = x + 3으로 한줄씩 수행을 하면 x = 4가 먼저 업데이트 되고 y = 4 + 3 = 7이 된다.
```python
n = 12
id(n) # 식별 번호
n += 1 # 13
id(n) # 식별 번호 변경
```
n이 int형 객체의 12를 참조하다가 int형 객체 13을 참조하는 것으로 변경된 것이므로 id가 달라진다.

파이썬의 대입
- 좌변에 변수 이름이 처음 나온 경우, 그 변수에 맞는 자료형으로 자동으로 선언
- 대입식은 값 자체가 아니라 참조하는 객체의 식별 번호를 대입
- 여러 변수에 여러 값을 동시에 대입할 수 있다

##### 리스트와 튜플 여러가지
- len() : 리스트나 튜플의 원소 개수
- min(), max() : 리스트나 튜플의 원소 중에서 최솟값, 최댓값 반환
- 빈 배열 판단
```python
if x:
    # x가 비어있지 않으면 True
else:
    # x가 비어있으면 False
```
- 배열의 대소 또는 등가 관계 판단
    - 맨 앞 원소부터 비교 하면서 원소의 값이 같으면 다음 원소를 비교
    - 어느 원소의 값이 크면 그 배열이 큰 것으로 판단
    - 원소 수가 다른 경우에는 원소 수가 많은 배열이 더 크다고 판단
    - 원소의 값을 비교하는 것이 우선순위
    ex) [50] < [3, 4] : 첫번째 원소를 비교하면서 50이 더 크기 때문에 False가 된다.

- 등가성(Equality)과 동일성(Identity)
    - 등가성 비교(==) : 좌변과 우변의 값이 같은지를 비교
    - 동일성 비교(is) : 값 뿐만 아니라 참조하고 있는 객체의 식별번호가 같은지 비교
    ```python
    lst1 = [1, 2, 3, 4, 5]
    lst2 = [1, 2, 3, 4, 5]
    print(lst1 == lst2) # True
    print(lst1 is lst2) # False
    ```
    - 리스트 두 개를 선언하여 서로 대입해도 원소를 복사하는 것은 아니다.
    ```python
    lst1 = [1, 2, 3, 4, 5]
    lst2 = lst1
    print(lst1 is lst2) # True
    ```


##### 자료구조의 개념
- <u>데이터 단위와 데이터 자체 사이의 물리적 또는 논리적인 관계</u>
- 컴퓨터에서 처리해야 하는 많은 데이터를 모아 효율적으로 관리하고 구조화
- 리스트와 튜플은 배열(Array)


#### 2. 배열이란?

- 배열 원소의 최댓값 구하기
```python
# a의 원소가 3개일 때 최댓값 구하기
maximum = a[0]
if a[1] > maximum:
    maximum = a[1]
if a[2] > maximum:
    maximum = a[2]
```
n개의 원소가 있다면 if문을 n-1번 실행해야 한다.

In [11]:
# 배열 a의 원소들 중 최댓값을 구하는 함수
def max_of(a):
    maximum = a[0]
    for i in range(1, len(a)):
        if a[i] > maximum:
            maximum = a[i]
    return maximum

if __name__ == '__main__':
    print('배열의 최댓값을 구합니다.')
    num = int(input('원소 수를 입력하세요.'))
    x = [None] * num
    
    for i in range(num):
        x[i] = int(input(f'x[{i}]값을 입력하세요.: '))

    print(f'최댓값은 {max_of(x)}입니다.')

배열의 최댓값을 구합니다.
최댓값은 56입니다.


##### 모듈(Module)
- 하나의 스크립트 프로그램
- `if __name__ == '__main__':`으로 작성
    - 스크립트 프로그램이 직접 실행될 때 변수 __name__은 '__main__'
    - 스크립트 프로그램이 import 될 때 변수 __name__은 원래 모듈 이름
- 모듈도 하나의 객체(Object)
- 모듈은 프로그램이 처음 import 되는 시점에서 그 모듈 객체가 생성되면서 초기화되는 구조
- `if __name__ == '__main__':`문은 filename.py를 직접 시작한 경우에만(즉, __name__과 '__main__'이 일치하는 경우) 참이 되어 실행이 된다.
- filename.py로 저장해두고 같은 경로에 있는 다른 파이썬 파일에서 
    ```python
    from filename import function_name
    ```
    으로 import해서 쓸 수 있다.

- 배열의 원솟값을 난수로 결정하기
    - 배열의 원소 수, 최솟값, 최댓값을 주고 조건에 맞는 배열 생성

In [1]:
import random
from max import max_of

print('난수의 최댓값을 구합니다.')
num = int(input('난수의 개수를 입력하세요: '))
lower = int(input('난수의 최솟값을 입력하세요.: '))
upper = int(input('난수의 최댓값을 입력하세요.: '))
x = [None] * num

for i in range(num):
    x[i] = random.randint(lower, upper)

print(f'{x}')
print(f'이 가운데 최댓값은 {max_of(x)}입니다.')

난수의 최댓값을 구합니다.
[41, 65, 85, 60, 48]
이 가운데 최댓값은 85입니다.


- 튜플, 문자열, 문자열 리스트의 최댓값 구하기

In [5]:
from max import max_of

t = (4, 7, 5.6, 2, 3.14, 1)
s = '안녕하세요'
a = ['DTS', 'AAC', 'FLAC', 'ZA']

print(f'{t}의 최댓값은 {max_of(t)}입니다.')
print(f'{s}의 최댓값은 {max_of(s)}입니다.') # 문자열의 문자 중 유니코드 값이 가장 큰 문자 출력
print(f'{a}의 최댓값은 {max_of(a)}입니다.') # 문자열을 원소로 갖는 리스트 중 최댓값은 사전 순으로 했을 때 가장 뒤에 나오는 문자열이 최댓값

(4, 7, 5.6, 2, 3.14, 1)의 최댓값은 7입니다.
안녕하세요의 최댓값은 하입니다.
['DTS', 'AAC', 'FLAC', 'ZA']의 최댓값은 ZA입니다.


- 배열 원소를 역순으로 정렬하기

    ex) [2, 5, 1, 3, 9, 6, 7] -> [7, 6, 9, 3, 1, 5, 2]

In [11]:
def reverse_array(a):
    n = len(a)
    for i in range(n // 2):
        a[i], a[n - 1 - i] = a[n - 1 - i], a[i]

if __name__ == '__main__':
    print('배열 원소를 역순으로 정렬합니다.')
    nx = int(input('원소 수를 입력하세요.: '))
    x = [None] * nx

    for i in range(nx):
        x[i] = int(input(f'x[{i}]의 값을 입력하세요.: '))

    reverse_array(x)

    print('배열의 원소를 역순으로 정렬했습니다.')
    for i in range(nx):
        print(f'x[{i}] = {x[i]}')

배열 원소를 역순으로 정렬합니다.
배열의 원소를 역순으로 정렬했습니다.
x[0] = 7
x[1] = 6
x[2] = 9
x[3] = 3
x[4] = 1
x[5] = 5
x[6] = 2


- 기수 변환하기(n진수 구하기)

In [4]:
# 10진수 정숫값을 입력받아 2~36진수로 변환하여 출력하기

def card_conv(x: int, r: int) -> str:
    '''
    정숫값 x를 r진수로 변환한 뒤 그 수를 문자열 타입으로 반환
    '''

    d = '' # 빈 문자열 생성
    dchar = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ' # 0 ~ 9 : 10개 + A ~ Z : 26 = 36개

    while x > 0: # x가 0보다 큰 경우에는 계속 반복
        d += dchar[x % r] # x 나누기 r의 나머지를 인덱스로 해서 dchar의 문자열 중 하나의 문자열을 d 문자열에 추가
        x //= r # x 나누기 r의 몫을 x에 할당해서 다시 반복
    
    return d[::-1] # 반환은 역순으로

if __name__ == '__main__':
    print('10진수를 n진수로 변환합니다.')

    while True: # 계속 반복
        while True: # 음수를 입력한 경우 계속 반복
            no = int(input("음이 아닌 정수를 입력하세요.: ")) 
            if no > 0: 
                break # 음이 아닌 정수가 입력되면 while문 탈출
        
        while True: # 2 ~ 36이 아닌 숫자가 입력됐을 경우 계속 반복
            cd = int(input("어떤 진수로 변환할까요?: "))
            if 2 <= cd <= 36:
                break # 2 ~ 36 사이의 값이 입력되면 while문 탈출
        
        print(f'{cd}진수로는 {card_conv(no, cd)}입니다.')

        retry = input("한 번 더 변환할까요?(Y : 예 / N : 아니요): ")

        if retry in {'N', 'n'}: 
            break # N, n이 입력됐을 경우 while문 탈출

10진수를 n진수로 변환합니다.
8진수로는 1030입니다.


In [8]:
# 진수 변환 과정을 보여주는 함수

def card_conv_verbose(x: int, r: int) -> str:
    '''
    정숫값 x를 r진수로 변환한 뒤 그 수를 문자열 타입으로 반환
    '''

    d = '' # 빈 문자열 생성
    dchar = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ' # 0 ~ 9 : 10개 + A ~ Z : 26 = 36개
    n = len(str(x))

    print(f'{r:2} | {x:{n}d}')

    while x > 0: # x가 0보다 큰 경우에는 계속 반복
        print('   +' + (n + 2) * '-')
        if x //r :
            print(f'{r:2} | {x // r:{n}d} ... {x % r}')
        else:
            print(f'     {x // r:{n}d} ... {x % r}')
        d += dchar[x % r] # x 나누기 r의 나머지를 인덱스로 해서 dchar의 문자열 중 하나의 문자열을 d 문자열에 추가
        x //= r # x 나누기 r의 몫을 x에 할당해서 다시 반복
    
    return d[::-1] # 반환은 역순으로

card_conv_verbose(29, 16)

16 | 29
   +----
16 |  1 ... 13
   +----
      0 ... 1


'1D'

- 소수(Prime Number) 나열하기

    - 소수의 조건 : **어떤 정수 n이 소수라면 n은 2부터 n-1까지의 어떤 정수로도 나누어 떨어지지 않는다.**
    - 합성수(Composite Number) : 소수가 아닌 수

In [7]:
# 1000 이하의 소수 나열하기

counter = 0 # 나누기를 수행한 횟수

for n in range(2, 1001): # 2 ~ 1000까지 숫자(1은 소수가 아님)
    for i in range(2, n): # 2부터 2, 2부터 3, 2부터 4, ..., 2부터 1000까지
        counter += 1 # 나누기 횟수 +1
        if n % i == 0: # n을 i로 나누었을 때 나머지가 0이면(= 나누어 떨어지면)
            break # 내부 for문을 탈출(한 번이라도 나누어 떨어지면 소수가 아니므로 탈출)
    else: # 내부 for문을 중간에 탈출한게 아니라면
        print(n) # n을 출력(소수이므로)

print(f'나눗셈을 실행한 횟수: {counter}')

2
3
5
7
11
13
17
19
23
29
31
37
41
43
47
53
59
61
67
71
73
79
83
89
97
101
103
107
109
113
127
131
137
139
149
151
157
163
167
173
179
181
191
193
197
199
211
223
227
229
233
239
241
251
257
263
269
271
277
281
283
293
307
311
313
317
331
337
347
349
353
359
367
373
379
383
389
397
401
409
419
421
431
433
439
443
449
457
461
463
467
479
487
491
499
503
509
521
523
541
547
557
563
569
571
577
587
593
599
601
607
613
617
619
631
641
643
647
653
659
661
673
677
683
691
701
709
719
727
733
739
743
751
757
761
769
773
787
797
809
811
821
823
827
829
839
853
857
859
863
877
881
883
887
907
911
919
929
937
941
947
953
967
971
977
983
991
997
나눗셈을 실행한 횟수: 78022


- 소수 찾기 알고리즘 개선 1
    - 개선점 : **n이 2와 3으로 나누어 떨어지지 않는다면 $2 \times 2 = 4$와 $2 \times 3 = 6$으로도 나누어 떨어지지 않는다.**
    - 소수의 조건 : **어떤 정수 n이 소수라면 n은 2부터 n-1까지 <u>어떤 소수</u>로도 나누어 떨어지지 않는다.**
    - ex) 7이 소수인지 판단하려면?
        - 7보다 작은 소수인 2, 3, 5로 나눗셈을 해서 나누어 떨어지지 않는 것을 보이면 된다.
        - $( \because )$ 2와 3으로 나누어 떨어지지 않으므로 4와 6은 해보지 않아도 된다.

In [8]:
# 1000 이하의 소수를 나열하기(알고리즘 개선 1)

counter = 0 # 나누기를 수행한 횟수
ptr = 0 # 먼저 찾은 소수의 개수
prime = [None] * 500 # prime : 소수를 저장할 리스트

prime[ptr] = 2 # prime 리스트의 첫번째 원소는 2
ptr += 1 # 2가 있으므로 찾은 소수의 개수는 1

for n in range(3, 1001, 2): # 3, 5, 7, 9, 11, 13, ... , 999(= 홀수만)
    for i in range(1, ptr): # 1부터 (찾은 소수의 개수-1)까지
        counter += 1 # 나누기 수행 횟수 +1
        if n % prime[i] == 0: # n이 prime[i]와 나누어 떨어지면
            break # for문을 탈출
    else: # 중간에 for문을 탈출하지 않았다면
        prime[ptr] = n # prime[ptr]에 n으로 값을 업데이트
        ptr += 1 # 소수가 추가됐으므로 소수의 개수도 +1

for i in range(ptr): # 0부터 ptr-1까지
    print(prime[i]) # prime 리스트의 원소를 하나씩 출력
print(f'나눗셈을 실행한 횟수: {counter}')

2
3
5
7
11
13
17
19
23
29
31
37
41
43
47
53
59
61
67
71
73
79
83
89
97
101
103
107
109
113
127
131
137
139
149
151
157
163
167
173
179
181
191
193
197
199
211
223
227
229
233
239
241
251
257
263
269
271
277
281
283
293
307
311
313
317
331
337
347
349
353
359
367
373
379
383
389
397
401
409
419
421
431
433
439
443
449
457
461
463
467
479
487
491
499
503
509
521
523
541
547
557
563
569
571
577
587
593
599
601
607
613
617
619
631
641
643
647
653
659
661
673
677
683
691
701
709
719
727
733
739
743
751
757
761
769
773
787
797
809
811
821
823
827
829
839
853
857
859
863
877
881
883
887
907
911
919
929
937
941
947
953
967
971
977
983
991
997
나눗셈을 실행한 횟수: 14622


- 소수 찾기 알고리즘 개선 2
    - ex) 넓이가 100인 직사각형의 가로, 세로 변의 길이의 경우의 수(단, 각 변의 길이는 정수이고 1은 제외)  
        $2 \times 50 \\ 4 \times 25 \\ 5 \times 20 \\ 10 \times 10 \\ 20 \times 5 \\ 25 \times 4 \\ 50 \times 2$
    - 어떤 임의의 정수 n을 직사각형의 넓이로 봤을 때, 직사각형 한 변의 길이만 나눗셈을 시도하고, 그 과정에서 한 번도 나누어 떨어지지 않으면 소수로 판단
    - 소수의 조건 : **어떤 정수 n이 소수라면 n은 <u>n의 제곱근($\sqrt {n}$) 이하의 소수</u>로 나누어 떨어지지 않는다.**

In [10]:
# 1000이하의 소수 나열하기(알고리즘 개선 2)

counter = 0 # 곱셈, 나눗셈 연산 횟수
ptr = 0 # 앞서 찾은 소수의 개수
prime = [None] * 500 # 소수를 저장할 리스트

prime[ptr] = 2 # 2는 소수
ptr += 1 # 소수의 개수 +1

prime[ptr] = 3 # 3은 소수
ptr += 1 # 소수의 개수 +1

for n in range(5, 1001, 2): # 5, 7, 9, 11, 13, 15, ..., 999까지(= 홀수)
    i = 1 # prime 리스트의 인덱스 prime[1] = 3
    while prime[i] * prime[i] <= n: # 소수의 제곱이 n보다 작거나 같은 동안 반복
        counter += 2 # 곱셈 한 번, 나눗셈 한 번 하므로 +2
        if n % prime[i] == 0: # n 나누기 prime[i]가 나누어 떨어지면
            break # while문을 탈출
        i += 1 # 나누어 떨어지지 않은 경우에만 +1
    else: # while문을 탈출하지 않았다면(= 끝까지 나누어 떨어지지 않았다면)
        prime[ptr] = n # n은 새로운 소수이므로 prime 리스트에 추가
        ptr += 1 # 소수의 개수 +1
        counter += 1 # break를 통해 while문을 탈출했다면 prime[i] * prime[i]의 곱셈은 연산 횟수에 추가되지 않으므로 나누기 연산 횟수 한 번만 추가

for i in range(ptr):
    print(prime[i])
print(f'곱셈과 나눗셈을 실행한 횟수: {counter}')

2
3
5
7
11
13
17
19
23
29
31
37
41
43
47
53
59
61
67
71
73
79
83
89
97
101
103
107
109
113
127
131
137
139
149
151
157
163
167
173
179
181
191
193
197
199
211
223
227
229
233
239
241
251
257
263
269
271
277
281
283
293
307
311
313
317
331
337
347
349
353
359
367
373
379
383
389
397
401
409
419
421
431
433
439
443
449
457
461
463
467
479
487
491
499
503
509
521
523
541
547
557
563
569
571
577
587
593
599
601
607
613
617
619
631
641
643
647
653
659
661
673
677
683
691
701
709
719
727
733
739
743
751
757
761
769
773
787
797
809
811
821
823
827
829
839
853
857
859
863
877
881
883
887
907
911
919
929
937
941
947
953
967
971
977
983
991
997
곱셈과 나눗셈을 실행한 횟수: 3774


- 첫번째 알고리즘의 계산 횟수 78,022에서, 한 번 개선을 한 알고리즘은 14,622, 또 한 번 개선한 알고리즘은 3,774로 연산 횟수가 점점 줄었다.
- 대신 첫번째 알고리즘보다 보다 많은 메모리를 요구한다.(먼저 찾은 소수를 저장하는 리스트)

    => <u>**빠른 알고리즘은 그만큼 많은 메모리를 요구한다.**</u>

- 얕은 복사(Shallow Copy)와 깊은 복수(Deep Copy)
    - 얕은 복사 : 객체가 갖는 멤버의 값을 새로운 객체로 복사할 때 객체가 참조 자료형의 멤버를 포함하는 경우. 참조값만 복사
    - 깊은 복사 : 참조값 뿐만 아니라 참조하는 객체 자체를 복사(원본이 업데이트되도 바뀌지 않음)

In [11]:
# 얕은 복사
x = [[1, 2, 3], [4, 5, 6]]
y = x.copy() # 얕은 복사
x[0][1] = 9
print(x)
print(y) # 복사한 y도 같이 수정됨

[[1, 9, 3], [4, 5, 6]]
[[1, 9, 3], [4, 5, 6]]


In [12]:
# 깊은 복사
import copy

x = [[1, 2, 3], [4, 5, 6]]
y = copy.deepcopy(x) # 깊은 복사
x[0][1] = 9
print(x)
print(y) # 복사된 y는 수정되지 않음

[[1, 9, 3], [4, 5, 6]]
[[1, 2, 3], [4, 5, 6]]
